Sensible defaults for Cilium Envoy Config

jrajahalme opened the pull request:
CiliumEnvoyConfig needs BPF NodePort to function properly in most cases. Enable this by default if CEC is enabled, and kubeproxy replacement is not explicitly disabled.
    
For this to work, helm chart is made to default to KPR=false starting on 1.14. Validation will now fail only if KPR=disabled is explicitly configured or if KPR option is not given and upgradeCompatibility is < 1.14, when KPR will default to a disabled or probe.

```release-note
BPF NodePort is now enabled by default if CiliumEnvoyConfig is configured.
```

jrajahalme committed:
daemon: Enable options needed for CiliumEnvoyConfig if enabled

CiliumEnvoyConfig (also used by Ingress and GatewayAPI controllers) needs
BPF NodePort to function properly in most cases. Enable BPF NodePort by
default if Envoy config is enabled, and kube proxy replacement is not
explicitly disabled.

For this to work, helm chart is made to default to KPR=false starting
on 1.14. Validation will now fail only if KPR=disabled is explicitly
configured or if KPR option is not given and upgradeCompatibility is <
1.14, when KPR will default to a disabled or probe.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

sayboras commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -642,6 +642,19 @@ enableK8sEventHandover: false
 # -- Enable CiliumEndpointSlice feature.
 enableCiliumEndpointSlice: false
 
+envoyConfig:
```
you need to put this block in values.yaml.tmp file as well ?

jrajahalme replied:
You tell me :-) Do I?

sayboras replied:
haha sorry about wrong question mark usage.

Recently, values.yaml.tmpl is the source of truth, and values.yaml is actually generated from values.yaml.tmpl file.

jrajahalme replied:
Done :-)

meyskens commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -62,6 +62,14 @@ func initKubeProxyReplacementOptions() error {
 		return nil
 	}
 
+	if option.Config.KubeProxyReplacement == option.KubeProxyReplacementPartial {
```
can this be merged into 1 if statement? instead of a nested if?

jrajahalme replied:
It can, but IMO this reads better in the context, which is formed with similar if statements.

gentoo-root commented on install/kubernetes/cilium/templates/validate.yaml:
```diff
@@ -54,12 +54,12 @@
   {{- if hasKey .Values "kubeProxyReplacement" }}
-    {{- if and (ne .Values.kubeProxyReplacement "partial") (ne .Values.kubeProxyReplacement "strict") }}
+    {{- if (eq .Values.kubeProxyReplacement "disabled") }}
       {{ fail "Ingress/Gateway API controller requires .Values.kubeProxyReplacement to be set to either 'partial' or 'strict'" }}
```
The error messages should also mention envoy now, shouldn't they?

jrajahalme replied:
Yes, changed messages to this form:
```
Ingress/Gateway API controller and EnvoyConfig require ...
```

gentoo-root commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -50,6 +50,14 @@
 
+{{- /* Default values when 1.14 was initially deployed */ -}}
+{{- if semverCompare ">=1.14" (default "1.14" .Values.upgradeCompatibility) -}}
+  {{- if or .Values.envoyConfig.enabled .Values.ingressController.enabled .Values.gatewayAPI.enabled }}
```
Ingress controller and gateway API now also set KPR=partial automatically after this change, right? Why are these features not mentioned in the commit message?

jrajahalme replied:
Fixed, thanks!

squeed commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -650,6 +650,19 @@ enableK8sEventHandover: false
 enableCiliumEndpointSlice: false
 
+envoyConfig:
+  # -- Enable envoy config CRD
```
nit: add the full Kind, to make it easier to grep for.

jrajahalme replied:
Done.

squeed commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -650,6 +650,19 @@ enableK8sEventHandover: false
+envoyConfig:
+  # -- Enable envoy config CRD
+  # Cilium Envoy Config CRD can also be implicitly enabled by other options.
+  enabled: false
```
suggestion: set this value to `~` so we are free to update the defaulting later.

jrajahalme replied:
No. We can't punt the defaulting decision to the agent as the helm chart needs to know if this is enabled or not so that it can create associated secrets namespace if this is enabled. Also, have similar definitions for `ingressController` and `gatewayAPI` that are used in the same way and it could be somewhat confusing if this was defined differently. Due to the interplay with these other options, we actually do not set the corresponding `--enable-envoy-config` to `false` when `envoyConfig.enabled` is `false`, which means that setting this as `~` (nil) vs. `false` has no functional difference, as we only ever test if this is `true` or not.

squeed replied:
From an end-user, I would expect to be able to set `foo.enabled=false` and have that disable foo. But `envoyConfig.enabled=false` doesn't disable `envoyConfig`. Rather, it sets it to "default".

I would expect a value of `~` to mean "please compute a reasonable default".

That said, if this is consistent with how we've exposed other variables, I'm not going to be picky.

squeed replied:
oh sorry, I forgot that this part got peeled off. Got my notification dates turned around.

squeed commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -650,6 +650,19 @@ enableK8sEventHandover: false
+  # -- SecretsNamespace is the namespace in which envoy SDS will retrieve secrets from.
+  secretsNamespace:
+    # -- Create secrets namespace for CiliumEnvoyConfig CRDs.
+    create: true
```
Should this also be `~` -- true if envoyConfig evaluates to true?

jrajahalme replied:
We only ever test if this is `true` or not, so setting this as nil serves no purpose and adds an unneeded dimension to understanding the chart. It is simpler if this is always a well defined boolean.

As far as I have understood, not setting the default value in the helm values and letting Cilium Agent pick the default that could then later be changed in a new agent version (using the same chart version) would be the benefit of using `~` (`nil`) as the value here. In this case this does not apply, as `create` here controls only Helm behavior as there is no underlying Cilium Agent option to set.

squeed replied:
Nil has two main uses:

1. When there is any logic to determine the default value. This allows us to determine the default programmatically, while still allowing the end-user to override in either direction.
2. When the default value is and should be empty, but we would like to document the presence of a variable. Having a nil value means the docblock is consumed by tooling.

It not related directly to the agent.

squeed commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -219,8 +219,11 @@ data:
 {{- end }}
 
-{{- if or .Values.ingressController.enabled .Values.gatewayAPI.enabled (and (hasKey .Values "loadBalancer") (eq .Values.loadBalancer.l7.backend "envoy")) }}
+{{- if or .Values.envoyConfig.enabled .Values.ingressController.enabled .Values.gatewayAPI.enabled (and (hasKey .Values "loadBalancer") (eq .Values.loadBalancer.l7.backend "envoy")) }}
```
Suggestion: evaluate `enable-envoy-config` once to a variable, then consume that when, e.g. determining whether or not to create the secrets bits.

jrajahalme replied:
Could you elaborate on this suggestion? The way I see it is the variable would have the value `.Values.envoyConfig.enabled` and the condition in the secret creation would use it but not become any simpler. This is due to the secrets namespace creation having other constraints:
- it can be turned off even if envoyConfig is enabled
- it should not be created if the same namespace was already created for ingressController or gatewayAPI

squeed reviewed (changes requested):
looks good, just some tweaks around defaulting.

qmonnet reviewed (commented):
Looks good!

> For this to work, helm chart is made to default to KPR=partial starting on 1.14 if Envoy config, Ingress controller, or Gateway API is
enabled.

This should probably be mentioned in Documentation/operations/upgrade.rst? Maybe the new Helm values as well (although I'm not sure we've documented the other additions)?

meyskens reviewed (approved):
LGTM

jrajahalme committed:
helm: Add configmap default values for Cilium 1.14

kubeProxyReplacement default changes to "false" for new installs and for
explicit installs with upgradeCompatibility >= 1.14.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
